_drafts/unittests-cpputest.md

```diff
@@ -0,0 +1,1251 @@
+
+{:.no_toc}
+
+### Life Before Unit Testing
+
+Here are a few examples that I've experienced in the past that were alleviated
+by the organization doubling down on unit testing our firmware.
+
+- You find testing on target hardware **slow** and **inconvenient**, especially
+  when multiple devices (e.g. a mobile phone) or prior setup (e.g. a factory
+  reset) is required
+- Bugs and regressions occur **repeatedly** in a single piece of software.
+- Deadlocks, HardFaults and Memory Leaks are the norm and have become accepted
+  (unit testing example for preventing deadlocks included below!).
+- The amount of time spent debugging and testing firmware is **60% or
+  more**[^1].
+- The first instinct when starting a new software module is to write a chunk of
+  code and **test on hardware**.
+- Testing a majority of the firmware before pushing to master is
+  **incomprehensible**.
+
+{:.no_toc}
+
+### Life After Unit Testing (Possibly)
+
+At a previous company, after scrapping most legacy code and writing new modules
+with 90%+ code coverage and through the use of TTD, this is what development
+felt like sometimes.
+
+- You write a new file, maybe an in-flash log storage module, and it works the
+  **first time** when flashed on the device (no better feeling).
+- Regressions are caught **immediately** when runnings tests locally or in CI.
+- Memory leaks are raised as errors in unit tests.
+- Testing a majority of the firmware only takes **a minute**.
+- Writing and debugging code on a plane is possible! No hardware required.
+
+## Disclaimers
+
+Unit tests in the embedded space is a controversial topics, so I want to clear
+set up expectations up front.
+
+1. This post covers how to test embedded **software**. Testing firmware drivers
+   and hardware is _very_ different and time is best spent writing functional
+   and integration tests that run on target to validate hardware components. As
+   soon as drivers are written and stable, switch to the unit test approaches
+   provided in this post.
+2. I do not suggest rewriting all of your code to accommodate unit tests, or
+   writing tests for the current code base, but I heavily suggest writing unit
+   tests for most **new** modules and heavily suggesting them in code reviews.
+3. Integration tests and on-target tests have their place. This infrastructure
+   is a **huge** time and money investment, and the tests run in minutes and
+   hours. Keep these to a minimum at first to ensure hardware stability, and
+   leave software stability to unit tests. If time allows, then build these
+   types of tests.
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest posts
+straight to your mailbox_
+
+## Framework-less Unit Tests
+
+It is very common to initially write unit tests using one-off `.c` files. Below
+is an example of a test that is commonly found in firmware projects or written
+by the author of a piece of firmware code.
+
+```c
```

Comment:
Let's split this example in two. 

Let's imagine we have a my_sum.c module which exposes the `my_sum` function below:
```
int my_sum(int a, int b) {
  return a + b;
}
```

We would write a separate C file to test our function, let's call it test_my_sum.c:
```
#include <assert.h>

int main(int argc, char *argv[]) {
  assert(2 == my_sum(1, 1));
  assert(-2 == my_sum(-1, -1));
  assert(0 == my_sum(0, 0));
  // ...
  return(0);
}
```